src/v/resource_mgmt/cpu_profiler.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/property.h"
+#include "seastarx.h"
+
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/backtrace.hh>
+
+#include <absl/container/node_hash_map.h>
+
+#include <chrono>
+#include <cstdint>
+
+namespace resources {
+
+class cpu_profiler : public ss::peering_sharded_service<cpu_profiler> {
+    // The maximum number of backtraces the cpu_profiler will be allowed
+    // to retain per shard.
+    static constexpr uint64_t number_of_backtraces_to_retain{256};
+
+public:
+    struct cpu_profiler_result {
+        ss::sstring user_backtrace;
+        size_t occurrences;
+
+        cpu_profiler_result(ss::sstring ub, size_t o)
+          : user_backtrace(std::move(ub))
+          , occurrences(o) {}
+    };
+
+    using cpu_profiler_results
+      = std::vector<std::pair<ss::shard_id, std::vector<cpu_profiler_result>>>;
+
+    cpu_profiler(
+      config::binding<bool>&&, config::binding<std::chrono::milliseconds>&&);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    ss::future<cpu_profiler_results>
+    results(std::optional<ss::shard_id> shard_id);
+
+    std::vector<cpu_profiler_result> shard_results() const;
+
+private:
+    // Used to poll seastar at set intervals to capture all samples
+    ss::timer<ss::lowres_clock> _query_timer;
+    ss::gate _gate;
+
+    // Configuration for seastar's cpu profiler
+    config::binding<bool> _enabled;
+    config::binding<std::chrono::milliseconds> _sample_rate;
+
+    // A buffer to copy sampled traces from seastar into so
+    // we're not reallocating every call.
+    std::vector<seastar::cpu_profiler_trace> _results_buffer;
+
+    struct backtrace_details {
+        uint64_t occurrences;
+    };
+    // This map will have at most `_cfg.number_of_backtraces_to_retain` entries.
+    // If a new backtrace needs to be added and there is no space for it then
+    // an existing backtrace will be selected and evicted from the map to make
+    // room.
+    absl::node_hash_map<ss::simple_backtrace, backtrace_details> _backtraces;
```

Comment:
My thought was that by aggregating identical traces ahead of time we'd be able to store more traces. In my initial testing it's not uncommon to see the same trace 20+ times so it didn't make sense to just store 20 identical traces.

Reply:
I see. Wondering is that still a representative sample then?

Reply:
Ah, nice catch, it's not obviously representative of anything.  Vaguely it's correlated with more recent activity.

I think I have two options here;
- Drop the hash_map entirely and store samples in an LRU without aggregating them ahead of time. This method would result in a sample set that directly represent what a shard was doing for the last N seconds where N is some function on the max size of the LRU.
- Switch to using reservoir sampling as the replacement policy of for the hash_map. This would result in a sample set that is representative of what the shard was doing since the start of RP's execution.

For the time being I'm doing to switch to the LRU. It's easier to implement and I believe people will find it more useful in the short term. We can always implement the second option later if the need arises.

Reply:
Yeah I think switch to LRU is fine.

I do think it's a good idea to aggregate the samples to store more samples, as you mention, and also to reduce the size of the output in the API (the format needs to include "count"). It can still be an LRU with 1 entry per sample, but the entries point to a shared backtrace + count for any samples that had identical backtraces. So it's logically an LRU but uses aggregation ("interning") as an optimization (and this optimization could extend to the output by only outputting identical backtraces once, with a count indicator or some kind of reference).